scapy/sendrecv.py

```diff
@@ -574,94 +561,150 @@ def srpflood(x, promisc=None, filter=None, iface=None, iface_hint=None, nofilter
+
+  >>> sniff(iface="eth0", prn=Packet.summary)
+
+  >>> sniff(iface=["eth0", "mon0"],
+  ...       prn=lambda pkt: "%s: %s" % (pkt.sniffed_on, pkt.summary()))
+
+  >>> sniff(iface={"eth0": "Ethernet", "mon0": "Wifi"},
+  ...       prn=lambda pkt: "%s: %s" % (pkt.sniffed_on, pkt.summary()))
+
     """
     c = 0
-    label = {}
-    sniff_sockets = []
+    sniff_sockets = {}  # socket: label dict
     if opened_socket is not None:
-        sniff_sockets = [opened_socket]
-    else:
-        if offline is None:
-            if L2socket is None:
-                L2socket = conf.L2listen
-            if isinstance(iface, list):
-                for i in iface:
-                    s = L2socket(type=ETH_P_ALL, iface=i, *arg, **karg)
-                    label[s] = i
-                    sniff_sockets.append(s)
-            else:
-                sniff_sockets = [L2socket(type=ETH_P_ALL, iface=iface, *arg,
-                                           **karg)]
+        if not isinstance(opened_socket, (dict, list)):
+            opened_socket = [opened_socket]
+        if isinstance(opened_socket, dict):
+            sniff_sockets.update((s, label)
+                                 for s, label in iteritems(opened_socket))
+        else:
+            sniff_sockets.update((s, "socket%d" % i)
+                                 for i, s in enumerate(opened_socket))
+    if offline is not None:
+        flt = karg.get('filter')
+        if not isinstance(offline, (dict, list)):
+            offline = [offline]
+        if isinstance(offline, dict):
+            sniff_sockets.update((PcapReader(
+                fname if flt is None else
+                tcpdump(fname, args=["-w", "-", flt], getfd=True)
+            ), label) for fname, label in iteritems(offline))
         else:
-            flt = karg.get('filter')
-            sniff_sockets = [PcapReader(
-                offline if flt is None else
-                tcpdump(offline, args=["-w", "-", flt], getfd=True)
-            )]
+            sniff_sockets.update((PcapReader(
+                fname if flt is None else
+                tcpdump(fname, args=["-w", "-", flt], getfd=True)
+            ), fname) for fname in offline)
+    if not sniff_sockets or iface is not None:
+        if L2socket is None:
+            L2socket = conf.L2listen
+        if not isinstance(iface, (dict, list)):
+            iface = [iface]
+        if isinstance(iface, dict):
+            sniff_sockets.update(
+                (L2socket(type=ETH_P_ALL, iface=ifname, *arg, **karg), iflabel)
+                for ifname, iflabel in iteritems(iface)
+            )
+        else:
+            sniff_sockets.update(
+                (L2socket(type=ETH_P_ALL, iface=ifname, *arg, **karg), ifname)
+                for ifname in iface
+            )
     lst = []
     if timeout is not None:
         stoptime = time.time()+timeout
     remain = None
+    read_allowed_exceptions = ()
+    if conf.use_bpf:
```

Comment:
Same as before, could we move this piece of code in a dedicated function ?

Reply:
See my answer to your previous comment.